### ISSUE TITLE
Exhaustiveness checker does not understand that a cast pattern rules out non-matching types

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#51986]
Consider the following code (adapted from `co19/LanguageFeatures/Patterns/Exhaustiveness/lifting_cast_pattern_A01_t01`):
```
int test1(Object obj) {
  switch (obj) {
    case int(isEven: true) as int:
      return 1;
    case int _:
      return 2;
  }
}

int test2(Object obj) => switch (obj) {
    int(isEven: true) as int => 1,
    int _ => 2
  };
```
In `test1`, flow analysis understands that after `case int(isEven: true) as int:`, the scrutinee must be an `int`, so it promotes the matched value to `int`, and then concludes that `case int _:` will always match.  So it considers the switch statement to be trivially exhaustive, and doesn't complain the fact that there's no `return` statement after the switch.

In `test2`, flow analysis assumes the switch expression is exhaustive and leaves it to the exhaustiveness checker to determine whether the switch is exhaustive.  The exhaustiveness checker says:
```
The type 'Object' is not exhaustively matched by the switch cases since it doesn't match 'Object()'.
Try adding a wildcard pattern or cases that match 'Object()'.
```
It would be nice if the exhaustiveness checker also understood that the switch was exhaustive.

Note that if the type of `obj` is changed to `int?` instead of `Object`, then the exhaustiveness checker understands that the switch is exhaustive.

CC @johnniwinther

**Comment by sgrekhov** [comment]
I'm afraid this issue is still not fixed. Just tested on the freshly built SDK with the fix above included. `test2()` from the example above is still not exhaustve on both CFE and analyzer. Please run
`co19/LanguageFeatures/Patterns/Exhaustiveness/lifting_cast_pattern_A01_t02` to verify

**Comment by eernstg** [comment]
This is a tricky case that puts the spotlight on differences between the flow analysis and the exhaustiveness analysis.

The exhaustiveness analysis does not have the ability to represent a space which is the complement of a type test (in this case: something that describes "every object which isn't an `int`"). So we can't produce a very good result for the first case, it is just considered to cover every `int` whose `isEven` evaluates to `true`. In particular, exhaustiveness does _not_ record any information such as "if `obj` is not an `int` then it is also covered by case 1".

Flow analysis, on the other hand, notes that the scrutinee has passed a test that confirms that it is an `int` if it ever reaches the second case, and this implies that it will not reach the end of the switch: If we don't get an exception then we reach a `return ...`.

So there's no error in `test1` saying that it might return null because it reaches the end.

I don't think we can expect to eliminate these differences between the exhaustiveness analysis and the flow analysis. This would imply that we would accept the current behavior as correct, even though it looks like an inconsistency that `test2` causes an error but `test1` does not.

**Comment by sgrekhov** [comment]
@eernstg thank you! I'll update the test. We also have
```dart
sealed class A {
  final int field;
  A(this.field);
}

class B extends A {
  B(int field) : super(field);
}

class C extends A {
  C(int field) : super(field);
}

test1(A a) {
  switch (a) { // No error. But what about B?
    case C() as C:
      return 0;
  }
}

test2(A a) {
  switch (a) { // Error. The type 'A' is not exhaustively matched by the switch cases since it doesn't match 'C()'. 
    case B() as C:
      return 0;
  }
}
```
Is this expected?

**Comment by eernstg** [comment]
`test1` is fine: If the type check `as C` throws then it is covered, and if the type check `as C` completes normally then it is guaranteed that matching the object pattern `C()` against the matched value will succeed.

In terms of the rules, we are covering the entire space union of the scrutinee (`M` in the spec) because the type `C` is a subset of the space union of the pattern `C()`.

`test2` _must_ be an error, because we have a counterexample: An instance of `C` will pass the type check `as C`, and then it will fail to pass the object pattern `B()`, and this shows that the switch expression isn't exhaustive. A sealed class is an always-exhaustive type, and hence we get an error about exhaustiveness.

In terms of the rules, we can only conclude that the single case covers all objects of type `B` (because some objects in the set of objects surviving `as C` will not match `B()` successfully), and hence it isn't exhaustive.

In other words: Yes, that is expected.

**Comment by sgrekhov** [comment]
Thank you. Closing this issue
